File: skills/current/omp-search/lib/parser.ts
import type { SearchSource } from "./models.ts";

export const ANSI_RE = /\x1b(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])/g;
export const OSC_RE = /\x1b\][^\x07]*(?:\x07|\x1b\\)/g;
export const HEADER_RE = /Web Search:\s*(?<provider>.+?)\s+(?<count>\d+)\s+sources?\b/;
export const SECTION_RE = /[-─]{3,}\s*(?<name>Answer|Sources|Metadata)\b/i;
export const SOURCE_RE =
  /^(?:[+*-]|[\u251c\u2514\u2500─]+\s*)\s*(?<title>.+?)\s+\((?<domain>[^();·]+?)(?:[;·]\s*(?<ageIn>[^()]+?))?\)(?:\s*[·;]\s*(?<ageOut>.+?))?$/;
export const MORE_LINES_RE = /(?:…|\.\.\.)\s*\d+\s+more\s+lines/;
export const SECRET_RE = /\bsk-[a-z0-9_-]{8,}\b/gi;
export const ASSIGNMENT_SECRET_RE =
  /(\b[A-Z][A-Z0-9_]*(?:API_KEY|TOKEN|SECRET)\b\s*[=:]\s*)\S+/gi;
export const AUTH_SECRET_RE = /(\bAuthorization\s*:\s*Bearer\s+)\S+/gi;

export function stripTerminalControls(value: string): string {
  let cleaned = value.replace(OSC_RE, "");
  cleaned = cleaned.replace(ANSI_RE, "");
  return cleaned.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
}

export function redact(value: string): string {
  let cleaned = value.replace(SECRET_RE, "<redacted>");
  cleaned = cleaned.replace(ASSIGNMENT_SECRET_RE, "$1<redacted>");
  return cleaned.replace(AUTH_SECRET_RE, "$1<redacted>");
}

export function frameContent(line: string): string {
  let content = line;
  if (content.includes("│")) {
    const firstIdx = content.indexOf("│");
    content = content.slice(firstIdx + 1);
    if (content.includes("│")) {
      const lastIdx = content.lastIndexOf("│");
      content = content.slice(0, lastIdx);
    }
  }
  return content.trimEnd();
}

export interface ParsedSearchOutput {
  readonly query: string;
  readonly provider: string;
  readonly answer: string;
  readonly sources: readonly SearchSource[];
  readonly truncated: boolean;
  readonly parsed: boolean;
  readonly cleanedRaw: string;
}

export function parseSearchOutput(
  raw: string,
  fallbackQuery: string
): ParsedSearchOutput {
  const cleaned = stripTerminalControls(raw);
  const lines = cleaned.split("\n");
  let provider: string | null = null;
  let actualQuery = fallbackQuery;
  const answerLines: string[] = [];
  const sources: SearchSource[] = [];
  let section: "answer" | "sources" | "metadata" | null = null;
  let started = false;

  for (const line of lines) {
    if (!started) {
      const headerMatch = line.match(HEADER_RE);
      const providerGroup = headerMatch?.groups?.["provider"];
      if (providerGroup) {
        provider = providerGroup.trim();
        started = true;
        continue;
      }
    }

    if (line.includes("---") || line.includes("───")) {
      const sectionMatch = line.match(SECTION_RE);
      const nameGroup = sectionMatch?.groups?.["name"];
      if (nameGroup) {
        section = nameGroup.toLowerCase() as
          | "answer"
          | "sources"
          | "metadata";
        started = true;
        continue;
      }
      if (section !== null) {
        break;
      }
    }

    const content = frameContent(line);
    const trimmed = content.trim();

    if (/^query:\s*/i.test(trimmed)) {
      actualQuery = trimmed.replace(/^query:\s*/i, "").trim() || fallbackQuery;
      continue;
    }

    if (section === "answer") {
      answerLines.push(content);
    } else if (section === "sources") {
      const match = trimmed.match(SOURCE_RE);
      const title = match?.groups?.["title"];
      const domain = match?.groups?.["domain"];
      const ageIn = match?.groups?.["ageIn"];
      const ageOut = match?.groups?.["ageOut"];
      const age = ageIn || ageOut;
      if (title && domain) {
        sources.push({
          title: title.trim(),
          domain: domain.trim(),
          age: age ? age.trim() : null,
        });
      }
    } else if (section === "metadata" && /^provider:\s*/i.test(trimmed)) {
      provider = trimmed.replace(/^provider:\s*/i, "").trim() || provider;
    }
  }

  while (answerLines.length > 0 && !answerLines[0]) {
    answerLines.shift();
  }
  while (answerLines.length > 0 && !answerLines[answerLines.length - 1]) {
    answerLines.pop();
  }

  const answer = answerLines.join("\n");
  const truncated = MORE_LINES_RE.test(cleaned);
  const parsed = Boolean(provider || answer || sources.length > 0);

  return {
    query: actualQuery,
    provider: provider || "unknown",
    answer,
    sources,
    truncated,
    parsed,
    cleanedRaw: cleaned,
  };
}

File: skills/current/omp-search/lib/parser.py
"""Parse OMP terminal output into structured fields."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from models import SearchSource

ANSI_RE = re.compile(r"\x1b(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])")
OSC_RE = re.compile(r"\x1b\][^\x07]*(?:\x07|\x1b\\)")
HEADER_RE = re.compile(r"Web Search:\s*(?P<provider>.+?)\s+(?P<count>\d+)\s+sources?\b")
SECTION_RE = re.compile(
    r"[-─]{3,}\s*(?P<name>Answer|Sources|Metadata)\b", re.IGNORECASE
)
SOURCE_RE = re.compile(
    r"^(?:[+*-]|[├└─]+\s*)\s*(?P<title>.+?)\s+"
    r"\((?P<domain>[^();·]+?)(?:[;·]\s*(?P<ageIn>[^()]+?))?\)"
    r"(?:\s*[·;]\s*(?P<ageOut>.+?))?$"
)
MORE_LINES_RE = re.compile(r"(?:…|\.\.\.)\s*\d+\s+more\s+lines")
SECRET_RE = re.compile(r"\bsk-[a-z0-9_-]{8,}\b", re.IGNORECASE)
ASSIGNMENT_SECRET_RE = re.compile(
    r"(\b[A-Z][A-Z0-9_]*(?:API_KEY|TOKEN|SECRET)\b\s*[=:]\s*)\S+", re.IGNORECASE
)
AUTH_SECRET_RE = re.compile(r"(\bAuthorization\s*:\s*Bearer\s+)\S+", re.IGNORECASE)


def strip_terminal_controls(value: str) -> str:
    """Remove ANSI/OSC escape sequences and normalize newlines."""
    cleaned = OSC_RE.sub("", value)
    cleaned = ANSI_RE.sub("", cleaned)
    return cleaned.replace("\r\n", "\n").replace("\r", "\n")


def redact(value: str) -> str:
    """Mask API keys, token assignments, and bearer credentials."""
    cleaned = SECRET_RE.sub("<redacted>", value)
    cleaned = ASSIGNMENT_SECRET_RE.sub(r"\1<redacted>", cleaned)
    return AUTH_SECRET_RE.sub(r"\1<redacted>", cleaned)


def frame_content(line: str) -> str:
    """Strip box-drawing frame borders and trailing whitespace."""
    content = line
    if "│" in content:
        first_idx = content.index("│")
        content = content[first_idx + 1 :]
        if "│" in content:
            last_idx = content.rindex("│")
            content = content[:last_idx]
    return content.rstrip()


@dataclass(frozen=True)
class ParsedSearchOutput:
    """Structured fields extracted from one OMP search transcript."""

    query: str
    provider: str
    answer: str
    sources: list[SearchSource]
    truncated: bool
    parsed: bool
    cleaned_raw: str


def parse_search_output(raw: str, fallback_query: str) -> ParsedSearchOutput:
    """Parse OMP search output into query, provider, answer, and sources."""
    cleaned = strip_terminal_controls(raw)
    lines = cleaned.split("\n")
    provider: str | None = None
    actual_query = fallback_query
    answer_lines: list[str] = []
    sources: list[SearchSource] = []
    section: str | None = None
    started = False

    for line in lines:
        if not started:
            header_match = HEADER_RE.search(line)
            provider_group = header_match.group("provider") if header_match else None
            if provider_group:
                provider = provider_group.strip()
                started = True
                continue

        if "---" in line or "───" in line:
            section_match = SECTION_RE.search(line)
            name_group = section_match.group("name") if section_match else None
            if name_group:
                section = name_group.lower()
                started = True
                continue
            if section is not None:
                break

        content = frame_content(line)
        trimmed = content.strip()

        if re.match(r"^query:\s*", trimmed, re.IGNORECASE):
            actual_query = (
                re.sub(r"^query:\s*", "", trimmed, flags=re.IGNORECASE).strip()
                or fallback_query
            )
            continue

        if section == "answer":
            answer_lines.append(content)
        elif section == "sources":
            match = SOURCE_RE.match(trimmed)
            title = match.group("title") if match else None
            domain = match.group("domain") if match else None
            age_in = match.group("ageIn") if match else None
            age_out = match.group("ageOut") if match else None
            age = age_in or age_out
            if title and domain:
                sources.append(
                    {
                        "title": title.strip(),
                        "domain": domain.strip(),
                        "age": age.strip() if age else None,
                    }
                )
        elif section == "metadata" and re.match(
            r"^provider:\s*", trimmed, re.IGNORECASE
        ):
            provider = (
                re.sub(r"^provider:\s*", "", trimmed, flags=re.IGNORECASE).strip()
                or provider
            )

    while answer_lines and not answer_lines[0]:
        answer_lines.pop(0)
    while answer_lines and not answer_lines[-1]:
        answer_lines.pop()

    answer = "\n".join(answer_lines)
    truncated = bool(MORE_LINES_RE.search(cleaned))
    parsed = bool(provider or answer or len(sources) > 0)

    return ParsedSearchOutput(
        query=actual_query,
        provider=provider or "unknown",
        answer=answer,
        sources=sources,
        truncated=truncated,
        parsed=parsed,
        cleaned_raw=cleaned,
    )
